File: Leetcode/Dynamic Programming/#91 Decode Ways.cpp
class Solution {
public:
    vector<int>dp;
    int topdown(string s, int n){
        if(dp[n]!=-1) return dp[n];
        if(s[n]=='0') return dp[n]=0;
        dp[n]=topdown(s,n+1);
        if(n<s.size()-1 && (s[n]=='1' || (s[n]=='2' && ('0'<=s[n+1] && s[n+1]<='6')))) dp[n]+=topdown(s,n+2);
        return dp[n];
    }
    int numDecodings(string s) {
        if(s=="0") return 0;
        dp=vector<int>(s.size()+1,-1);
        dp[s.size()]=1;
        return topdown(s,0);
    
    }
};

File: Dynamic Programming/#91. Decode Ways.py
'''
TC: O(n)
SC: O(n)
'''
class Solution:
    def numDecodings(self, s: str) -> int:
        '''
        f(i) = # of decode ways of s[i:]
        [Base Case]
        f(i) = 1, if i >= n
        f(i) = f(i + 1) + f(i + 2), if i < n - 1 && 1 <= s[i:i+2] <= 26
        f(i) = f(i + 2) if > 26
        '''
        self.n = len(s)
        self.dp = [-1 for _ in range(self.n)]
        return self.topdown(s, 0)


    def topdown(self, s, i):
        if i >= self.n:
            return 1

        if s[i] == '0':
            return 0

        if self.dp[i] != -1:
            return self.dp[i]

        if i < self.n - 1 and int(s[i:i+2]) <= 26:
            self.dp[i] = self.topdown(s, i + 1) + self.topdown(s, i + 2)
        else:
            self.dp[i] = self.topdown(s, i + 1)

        return self.dp[i]
        